### PR TITLE
feat: Separate DrivesManager(s) from UI

### PR DESCRIPTION
**Pull request by lukeblevins**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Might help with #2568 
- Definitely helps with general codebase quality

**Details of Changes**
Add details of changes here.
- Reorganized code to separate DrivesManager implementation(s) from UI [sidebar] related constructs.
- Did the same for BulkObservableCollection with its thumbnail setter internals (moved to layout modes)
- ~~Drop ARM from sln file~~ _(Will be done in a future pull request)_
- **Requires extensive testing of drive enumeration reliability**

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] ~~Tested the changes for accessibility~~

### CODE REVIEW

File: Files/UserControls/Widgets/DrivesWidget.xaml [hunk 1]
```diff
@@ -178,11 +179,10 @@
                                 <Image
                                     Grid.RowSpan="3"
                                     Grid.Column="0"
+                                    x:Name="DriveThumbnailImage"
                                     Width="40"
                                     Height="40"
-                                    HorizontalAlignment="Center"
-                                    x:Phase="1"
```
Comment (yaichenbaum) [inline Files/UserControls/Widgets/DrivesWidget.xaml:184]: Shouldn't we still use `x:Phase`?

**Comment by yaichenbaum** [comment]
@duke7553 can you see if this works in release mode?

File: Files/UserControls/Widgets/DrivesWidget.xaml (hunk 1 above)
Comment (lukeblevins) [reply Files/UserControls/Widgets/DrivesWidget.xaml:184]: @yair100 Yes. I've added it back.

**Comment by lukeblevins** [comment]
@yair100 I also reverted the configuration changes to the sln file. We can always drop ARM from the sln file in a subsequent PR. Hopefully, release builds run now.

File: Files/UserControls/Widgets/DrivesWidget.xaml (hunk 1 above)
Comment (lukeblevins) [reply Files/UserControls/Widgets/DrivesWidget.xaml:184]: @yair100 Actually, I remembered why it was correctly removed. x:Phase must be used in conjunction with x:Bind.

**Review by yair100 — DISMISSED** [review]

File: Files/Filesystem/Drives.cs [hunk 2]
```diff
@@ -488,12 +371,12 @@ public async Task HandleWin32DriveEvent(DeviceEvent eventType, string deviceId)
                     DriveItem matchingDriveEjected = Drives.FirstOrDefault(x => x.DeviceID == deviceId);
                     if (rootModified && matchingDriveEjected != null)
                     {
-                        _ = CoreApplication.MainView.DispatcherQueue.EnqueueAsync(() =>
-                        {
+                        //_ = CoreApplication.MainView.DispatcherQueue.EnqueueAsync(() =>
+                        //{
                             matchingDriveEjected.Root = rootModified.Result;
                             matchingDriveEjected.Text = rootModified.Result.DisplayName;
```
Comment (gave92) [inline Files/Filesystem/Drives.cs:377]: Aren't these bound to UI? Does it work without `EnqueueAsync`?

File: Files/DataModels/NavigationControlItems/DriveItem.cs [hunk 3]
```diff
@@ -139,26 +135,21 @@ public static async Task<DriveItem> CreateFromPropertiesAsync(StorageFolder root
...
+            {
+                item.ThumbnailData = await imageStream.ToByteArrayAsync();
+            }
+
             item.Text = root.DisplayName;
             item.Type = type;
             item.Path = string.IsNullOrEmpty(root.Path) ? $"\\\\?\\{root.Name}\\" : root.Path;
             item.DeviceID = deviceId;
             item.Root = root;
-            await CoreApplication.MainView.DispatcherQueue.EnqueueAsync(() => item.UpdatePropertiesAsync());
```
Comment (gave92) [inline Files/DataModels/NavigationControlItems/DriveItem.cs:148]: Aren't MaxSpace, FreeSpace, etc bound to UI? Does it work without `EnqueueAsync`?

File: Files/Filesystem/WSLDistroManager.cs [hunk 4]
```diff
@@ -1,140 +1,86 @@
...
-        }
-
-        private async void EnumerateDrivesAsync(CoreApplicationView sender, Windows.ApplicationModel.Activation.IActivatedEventArgs args)
-        {
-            await SyncSideBarItemsUI();
-            CoreApplication.MainView.Activated -= EnumerateDrivesAsync;
-        }
+                if (App.AppSettings.ShowWslSection && WslSection == null)
+                {
+                    WslSection = new LocationItem()
```
Comment (gave92) [inline Files/Filesystem/WSLDistroManager.cs:31]: Why is WSL manager different from the others? It seems to be the only onw to create it's own sidebar section

File: Files/Views/MainPage.xaml.cs [hunk 5]
```diff
@@ -295,6 +560,14 @@ protected override void OnNavigatingFrom(NavigatingCancelEventArgs e)
...
                 SidebarControl.SidebarItemNewPaneInvoked -= SidebarControl_SidebarItemNewPaneInvoked;
             }
+            App.DrivesManager.RefreshCompleted -= DrivesManager_RefreshCompleted;
+            App.DrivesManager.RemoveDrivesSidebarSection -= DrivesManager_RemoveDrivesSidebarSection;
+            App.CloudDrivesManager.RefreshCompleted -= CloudDrivesManager_RefreshCompleted;
+            App.CloudDrivesManager.RemoveCloudDrivesSidebarSection -= CloudDrivesManager_RemoveCloudDrivesSidebarSection;
+            App.NetworkDrivesManager.RefreshCompleted -= NetworkDrivesManager_RefreshCompleted;
+            App.NetworkDrivesManager.RemoveNetworkDrivesSidebarSection -= NetworkDrivesManager_RemoveNetworkDrivesSidebarSection;
+            App.WSLDistroManager.RefreshCompleted -= WSLDistroManager_RefreshCompleted;
+            App.WSLDistroManager.RemoveWslSidebarSection -= WSLDistroManager_RemoveWslSidebarSection;
```
Comment (gave92) [inline Files/Views/MainPage.xaml.cs:570]: Planning to give the same treatment to LibraryManager?

**Comment by gave92** [comment]
Hiding/Showing a sidebar section breaks havoc on the icons.. possibly because of my changes 😅

File: Files/ViewModels/ItemViewModel.cs [hunk 6]
```diff
@@ -717,10 +717,14 @@ public async Task ReloadItemGroupHeaderImagesAsync()
                     foreach (var gp in FilesAndFolders.GroupedCollection)
                     {
                         var img = await GetItemTypeGroupIcon(gp.FirstOrDefault());
-                        await CoreApplication.MainView.DispatcherQueue.EnqueueAsync(() =>
+                        if (img is byte[])
                         {
-                            gp.Model.ImageSource = img;
-                        }, Windows.System.DispatcherQueuePriority.Low);
+                            gp.Model.ImageBytes = img as byte[];
```
Comment (gave92) [inline Files/ViewModels/ItemViewModel.cs:722]: Bound to UI Tag, crashes with wrong thread exception when list is grouped by file type.

File: Files/Helpers/ItemListDisplayHelpers/GroupedCollection.cs [hunk 7]
```diff
@@ -129,12 +128,20 @@ public bool ShowCountTextBelow
...
         {
-            get => imageSource;
-            set => SetPropertyWithUpdateDelay(ref imageSource, value);
+            get => imageBytes;
+            set => SetPropertyWithUpdateDelay(ref imageBytes, value);
+        }
+
+        private Uri imageUri;
+
+        public Uri ImageUri
```
Comment (gave92) [inline Files/Helpers/ItemListDisplayHelpers/GroupedCollection.cs:141]: This should probably be referenced somewhere

File: Files/Views/MainPage.xaml.cs (hunk 5 above)
Comment (gave92) [reply Files/Views/MainPage.xaml.cs]: Guess the same should be done for SidebarPinnedModel in the future.

File: Files/Filesystem/WSLDistroManager.cs (hunk 4 above)
Comment (duke7553) [reply Files/Filesystem/WSLDistroManager.cs]: @gave92 WslDistroManager used to be a lot worse before the work with this PR. Oh, and thanks for improving upon that :)

Comment (gave92) [reply Files/Filesystem/WSLDistroManager.cs]: Np, indeed there wasn't much left to do for WslDistroManager 👍

File: Files/ViewModels/ItemViewModel.cs (hunk 6 above)
Comment (duke7553) [reply Files/ViewModels/ItemViewModel.cs:722]: @gave92 Actually, do we support icons for the header groups? If not, then I'd like to remove this altogether.

File: Files/DataModels/NavigationControlItems/DriveItem.cs (hunk 3 above)
Comment (duke7553) [reply Files/DataModels/NavigationControlItems/DriveItem.cs:148]: @gave92 I always suspected there wouldn't be any threading issue even if the property is bound to UI elsewhere. I don't think property change notifications are dependent on the UI thread, but I could be wrong.

This is absolutely something worth verifying.

File: Files/ViewModels/ItemViewModel.cs (hunk 6 above)
Comment (gave92) [reply Files/ViewModels/ItemViewModel.cs:722]: Yes, if you group by file type and click on a header to go to the zoomed out view

**Comment by likeremix715** [comment]
Likeremix715@gmail.com

**Comment by d2dyno1** [comment]
@duke7553 Can you share an update on this?

**Comment by yaichenbaum** [comment]
> @duke7553 Can you share an update on this?

Let's give Luke time to get back to this, he's currently in college and working on top of that so his time to work on Files is pretty limited.

**Comment by gave92** [comment]
Aargh sorry didn't mean to delete the branch 🤦‍♂️

**Comment by BanCrash** [comment]
@gave92 then you can restore it, right?

image

**Comment by gave92** [comment]
Aha thanks, it doesn't offer that option from the phone app xD

**Pull request closed without merging**